azurelinuxagent/ga/agent_update.py

```diff
@@ -0,0 +1,577 @@
+import glob
+import json
+import os
+import shutil
+import time
+
+from azurelinuxagent.common import conf, logger
+from azurelinuxagent.common.event import add_event, WALAEventOperation
+from azurelinuxagent.common.exception import AgentUpgradeExitException, UpdateError
+from azurelinuxagent.common.future import ustr
+from azurelinuxagent.common.logger import LogLevel
+from azurelinuxagent.common.protocol.extensions_goal_state import GoalStateSource
+from azurelinuxagent.common.protocol.restapi import VMAgentUpdateStatuses, VMAgentUpdateStatus
+from azurelinuxagent.common.utils import fileutil, textutil
+from azurelinuxagent.common.utils.flexible_version import FlexibleVersion
+from azurelinuxagent.common.version import CURRENT_VERSION, AGENT_NAME, AGENT_DIR_PATTERN
+from azurelinuxagent.ga.exthandlers import HandlerManifest
+
+AGENT_ERROR_FILE = "error.json"  # File name for agent error record
+AGENT_MANIFEST_FILE = "HandlerManifest.json"
+MAX_FAILURE = 3  # Max failure allowed for agent before declare bad agent
+
+
+def get_agent_update_handler(protocol):
+    return AgentUpdateHandler(protocol)
+
+
+class PersistentUpdateHandlerState(object):
+    """
+    This class is primarily used to maintain the in-memory persistent state for the agent updates.
+    This state will be persisted throughout the current service run and might be modified by external classes.
+    """
+    report_error = None
+
+
+class AgentUpdateHandler(object):
+
+    def __init__(self, protocol):
+        self.__protocol = protocol
+        self.__ga_family = conf.get_autoupdate_gafamily()
+        self.__autoupdate_enabled = conf.get_autoupdate_enabled()
+        self.__gs_id = self.protocol.get_goal_state().incarnation
```

Comment:
not necessarily we update on extension goal state id changes, that could fail due to transient issue. Next attempt we could update without new goal state too.

`also, it seems like it is never updated`
what are you referring to?

Reply:
self.__gs_id seems to be initialized only in init and never updated
 
 there can be changes in the goal state without changes in the incarnation

Reply:
Still, I didn't get it. why do we need to update. At this point last goal state already updated and should have latest incarnation number. But we don't the fetch the goal state if incarnation changes. So, we update and get the requested version based on last updated goal state right. 

Since agent update depends on requested version and it's coming from extension goal state. I see your point of adding Id instead of incarnation.